nix/flake.nix

```diff
@@ -0,0 +1,148 @@
+{
+  description = "A GPU-accelerated cross-platform terminal emulator and multiplexer written by @wez and implemented in Rust";
+
+  inputs = {
+    nixpkgs.url = "github:NixOS/nixpkgs/nixpkgs-unstable";
+    flake-utils.url = "github:numtide/flake-utils";
+    rust-overlay = {
+      url = "github:oxalica/rust-overlay";
+      inputs = {
+        nixpkgs.follows = "nixpkgs";
+        flake-utils.follows = "flake-utils";
+      };
+    };
```

Comment:
I'd advise keeping rust-overlay since the MSRV of wezterm can shift out of sync with the nixpkgs version. This has happened twice over the last 10 months on my NUR.

Reply:
Good point, then I agree it seems worth it in this case :)

Reply:
@ThinkChaos  I'm interested in removing `flake-utils`, and I've read that article before and changed my own system flake to not use it anymore. I'm not 100% clear on what this would look like for this kind of project though. Is there an example I can look somewhere, or another repo, where something comparable to what I'm doing here is shown?

Reply:
Even if you remove it, it'll still be an input because `rust-overlay` uses it. But if you prefer

But the example in the [Do we really need flake-utils?](https://ayats.org/blog/no-flake-utils/#do-we-really-need-flake-utils) section is pretty close. The idea is to move the equivalent of `eachDefaultSystem` into the output attrset so each attr calls it when needed:
`packages = forAllSystems (pkgs: { ... })` instead of `output = { ... }: eachDefaultSystem (pkgs: { packages = ...; })`.

I don't have any flake I can link to (most my stuff is on localhost ATM), but I usually use this function:
```nix
  eachSystem = f:
    lib.genAttrs
      lib.systems.flakeExposed
      (system: f system inputs.nixpkgs.legacyPackages.${system});
```

And then write `package = eachSystem (system: pkgs: { ... });`.

Anyways if you want to resolve this as is it's fine by me since `flake-utils` will be an input as long as we use `rust-overlay`, so up to you if you want to change it now. That info can always be used as a reference later or in other contexts :)

Reply:
Ah, I understand. I was a bit stuck on 2 question:

1. Where does the `system` come from inside `forAllSystems`? In your example you just pass it in, which the blog post doesn't do. I suppose the blog assumes that you'll do that if you need it :sweat_smile:
2. Wouldn't there be a lot of duplication if I had to repeat `buildInputs` and `nativeBuildInputs` inside `packages.${system}.default` and `devShell.${system}`? I suppose it would have been possible to just have a `makeBuildInputs` function and a `makeNativeBuildInputs` function, which is the called with `pkgs`, but that doesn't feel clean and I haven't ever seen that anywhere.
I suppose this one is solved by having an extra file for the package, using `callPackage` and then using `inputsFrom` as well as `packages` in the devshell to add extra packages just for the devshell.

Since we have the dependency anyway and this way of structuring the flake still seems to be very common (even if it's technically not the best way), I suppose I'll keep it like this for now. I'll move the package definition into a separate file (`default.nix`? `wezterm.nix`? I don't know what the convention is there) and then look how much more work would it be to do this.

Reply:
1. `system` is a function argument from [`lib.genAttrs`](https://noogle.dev/f/lib/genAttrs) in this case, `lib.systems.flakeExposed` is just a predefined list containing supported targets for Nix. I'm actually not a fan of using flakeExposed since it includes stuff like powerpc64le and 32bit, which often just don't work, but still show up in `nix flake show`.
2. `inputsFrom = pkgs.callPackage ./nix {};` or similar is the correct approach here IMO.

If you really want to avoid bringing in flake-utils, you could also use [fenix](https://github.com/nix-community/fenix) instead; As long as you don't use Rust-Analyzer from there, you can even set `inputs.rust-analyzer.follows = ""` there.
However, since this is intended as a nightly/development flake; I'd urge you to *not optimize for flake size* obsessively, and rather include the needed tools for smooth development.

Reply:
I understand. I'm always looking for reduced complexity, and in this case it seems like the more minimal solution would also be slightly more complex 😄 I'll settle with moving the package to a file, and keep `flake-utils`. Thanks for all the explanations!